File: dairytrack-api/MilkProduction/app/models/farmer.py
# FILE: farmer.py
from app import db
from bcrypt import hashpw, gensalt, checkpw

class Farmer(db.Model):
    __tablename__ = 'farmers'

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    email = db.Column(db.String(100), unique=True, nullable=False)
    first_name = db.Column(db.String(50))
    last_name = db.Column(db.String(50))
    birth_date = db.Column(db.Date)
    contact = db.Column(db.String(15))
    religion = db.Column(db.String(50))
    address = db.Column(db.Text)
    gender = db.Column(db.String(10))
    total_cattle = db.Column(db.Integer, default=0)
    join_date = db.Column(db.Date)
    status = db.Column(db.String(20))
    role = db.Column(db.String(20), default='farmer')
    password = db.Column(db.String(128), nullable=False)
    created_at = db.Column(db.DateTime, default=db.func.current_timestamp())
    updated_at = db.Column(db.DateTime, default=db.func.current_timestamp(), onupdate=db.func.current_timestamp())

    def to_dict(self):
        return {
            'id': self.id,
            'email': self.email,
            'first_name': self.first_name,
            'last_name': self.last_name,
            'birth_date': self.birth_date.strftime('%Y-%m-%d') if self.birth_date else None,
            'contact': self.contact,
            'religion': self.religion,
            'address': self.address,
            'gender': self.gender,
            'total_cattle': self.total_cattle,
            'join_date': self.join_date.strftime('%Y-%m-%d') if self.join_date else None,
            'status': self.status,
            'role': self.role,
            'created_at': self.created_at.strftime('%Y-%m-%d %H:%M:%S'),
            'updated_at': self.updated_at.strftime('%Y-%m-%d %H:%M:%S') if self.updated_at else None
        }

    def set_password(self, password):
        self.password = hashpw(password.encode('utf-8'), gensalt()).decode('utf-8')

    def check_password(self, password):
        return checkpw(password.encode('utf-8'), self.password.encode('utf-8'))

    def __repr__(self):
        return f"Farmer('{self.first_name} {self.last_name}', '{self.birth_date}')"

File: dairytrack-mobile/lib/model/peternakan/farmer.dart
class Peternak {
  int? id; // Password menjadi opsional

  final String firstName;
  final String lastName;
  final String address;
  final String contact;
  final String email;
  final String gender;
  final String religion;
  final String role;
  final String status;
  int totalCattle = 0; // Total cattle default 0
  final DateTime birthDate;
  final DateTime join_date;
  final String? password; // Password menjadi opsional
  final DateTime createdAt;
  final DateTime updatedAt;

  Peternak({
    this.id, // Tidak lagi required

    required this.firstName,
    required this.lastName,
    required this.address,
    required this.contact,
    required this.email,
    required this.gender,
    required this.religion,
    required this.role,
    required this.status,
    this.totalCattle = 0,
    required this.birthDate,
    this.password, // Tidak lagi required
    required this.join_date,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Peternak.fromJson(Map<String, dynamic> json) {
    return Peternak(
      id: json['id'], // Bisa null

      firstName: json['first_name'] ?? '',
      lastName: json['last_name'] ?? '',
      address: json['address'] ?? '',
      contact: json['contact'] ?? '',
      email: json['email'] ?? '',
      gender: json['gender'] ?? '',
      religion: json['religion'] ?? '',
      role: json['role'] ?? '',
      status: json['status'] ?? '',
      totalCattle:
          json['total_cattle'] != null ? json['total_cattle'] as int : 0,
      birthDate: json['birth_date'] != null
          ? DateTime.parse(json['birth_date'])
          : DateTime.now(),
      password: json['password'], // Bisa null
      join_date: json['join_date'] != null
          ? DateTime.parse(json['join_date'])
          : DateTime.now(),
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id, // Hanya tambahkan jika tidak null
      'first_name': firstName,
      'last_name': lastName,
      'address': address,
      'contact': contact,
      'email': email,
      'gender': gender,
      'religion': religion,
      'role': role,
      'status': status,
      'total_cattle': totalCattle > 0 ? totalCattle : null,
      'birth_date': birthDate.toIso8601String(),
      if (password != null)
        'password': password, // Hanya tambahkan jika tidak null
      'join_date': join_date.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}
